packages/vite/src/node/ssr/ssrModuleLoader.ts

```diff
@@ -263,19 +263,23 @@ async function nodeImport(
 
   try {
     const mod = await dynamicImport(url)
-    return proxyESM(id, mod)
+    return proxyESM(mod)
   } finally {
     unhookNodeResolve()
   }
 }
 
 // rollup-style default import interop for cjs
-function proxyESM(id: string, mod: any) {
-  const defaultExport = mod.__esModule
-    ? mod.default
-    : mod.default
-    ? mod.default
-    : mod
+function proxyESM(mod: any) {
+  let defaultExport = mod
+  const visited = new WeakSet()
+  while ('default' in defaultExport && defaultExport.__esModule) {
+    if (visited.has(defaultExport)) {
+      break
+    }
+    visited.add(defaultExport)
+    defaultExport = defaultExport.default
+  }
```

Comment:
Should we also keep the second condition? When `__esModule` isn't present, we are still checking for `mod.default` and unwrapping it in the prev code (we really need more SSR deps tests)

Maybe something like this could work? (untested, unformatted, just to give you the idea)
```suggestion
function unwrapDefaultESM(mod: any) {
  const visited = new WeakSet()
	while (!visited.has(mod) && 'default' in mod && mod.__esModule) {
	  visited.add(mod)
	  mod = mod.default
  }
  return mod
}
  
function proxyESM(mod: any) {
  const defaultExport = mod.__esModule
    ? unwrapDefaultESM(mod.default)
    : mod.default
    ? mod.default
    : mod
```

Also for reference, do you have any links or explanations for why this is happening? Maybe we are missing an unwrapping somewhere else (ping @sodatea, @aleclarson so you are aware of this fix)

Reply:
I think you could write the last lines in the suggestion as `mod.default ? mod.default : mod` as `mod.default ?? mod`

Is there a reason to use `WeakSet` in either case? In both cases it looks like the `WeakSet` is created within a short-lived function and if you used just a plain `Set` it would quickly be garbage collected.